### PR TITLE
WinPR::Clipboard: Refactor synthetic file

### PR DESCRIPTION
**Pull request by KangLin**

### CODE REVIEW

File: winpr/libwinpr/clipboard/synthetic_file.c [hunk 1]
```diff
@@ -30,17 +30,12 @@
...
 #include <errno.h>
 
-#include <dirent.h>
-#include <fcntl.h>
-#include <unistd.h>
-#include <sys/types.h>
-#include <sys/stat.h>
+#ifdef _WIN32
+#include <winpr/wtypes.h>
+typedef SSIZE_T ssize_t;
```
Comment (akallabeth) [inline winpr/libwinpr/clipboard/synthetic_file.c:37]: this is wrong, newer windows versions define this type

File: winpr/libwinpr/clipboard/synthetic_file.c [hunk 2]
```diff
@@ -404,24 +365,24 @@ static BOOL add_file_to_list(wClipboard* clipboard, const char* local_name,
...
 {
-	const char* c = name;
-	const char* last_name = name;
+	const WCHAR* c = name;
+	const WCHAR* last_name = name;
 
 	while (*c++)
 	{
-		if (*c == '/')
+		if (*c == L'/')
```
Comment (akallabeth) [inline winpr/libwinpr/clipboard/synthetic_file.c:375]: `Be careful with `L'somestring'`, windows understands `UCS-2/UTF16` most other systems `UCS4/UTF32`

File: winpr/libwinpr/clipboard/synthetic_file.c [hunk 3]
```diff
@@ -307,91 +286,73 @@ static BOOL add_directory_entry_to_list(wClipboard* clipboard, const char* local
...
 	BOOL result = FALSE;
-	DIR* dirp = NULL;
-	WLog_VRB(TAG, "adding directory: %s", local_name);
-	dirp = opendir(local_name);
+	WCHAR namebuf[MAX_PATH];
+	HANDLE hFind = NULL;
+	WIN32_FIND_DATAW FindData;
 
-	if (!dirp)
+	swprintf(namebuf, L"%s/*.*", local_name);
```
Comment (akallabeth) [inline winpr/libwinpr/clipboard/synthetic_file.c:325]: `Be careful with `L'somestring'`, windows understands `UCS-2/UTF16` most other systems `UCS4/UTF32` so the resulting string will be broken on many systems

File: winpr/libwinpr/clipboard/synthetic_file.c [hunk 4]
```diff
@@ -269,34 +247,35 @@ static WCHAR* concat_remote_name(const WCHAR* dir, const WCHAR* file)
...
 	BOOL result = FALSE;
-	char* local_name = NULL;
+	WCHAR* local_name = NULL;
 	WCHAR* remote_name = NULL;
 	WCHAR* remote_base_name = NULL;
 
 	/* Skip special directory entries. */
-	if ((strcmp(entry->d_name, ".") == 0) || (strcmp(entry->d_name, "..") == 0))
+
+	if ((_wcscmp(pFileData->cFileName, L".") == 0) || (_wcscmp(pFileData->cFileName, L"..") == 0))
```
Comment (akallabeth) [inline winpr/libwinpr/clipboard/synthetic_file.c:269]: `Be careful with `L'somestring'`, windows understands `UCS-2/UTF16` most other systems `UCS4/UTF32`

File: winpr/libwinpr/clipboard/synthetic_file.c [hunk 5]
```diff
@@ -269,34 +247,35 @@ static WCHAR* concat_remote_name(const WCHAR* dir, const WCHAR* file)
 		return NULL;
 
 	memcpy(buffer, dir, len_dir * sizeof(WCHAR));
-	buffer[len_dir] = L'\\';
+	buffer[len_dir] = L'/';
```
Comment (akallabeth) [inline winpr/libwinpr/clipboard/synthetic_file.c:250]: `Be careful with `L'somestring'`, windows understands `UCS-2/UTF16` most other systems `UCS4/UTF32`

**Review by akallabeth — CHANGES_REQUESTED** [review]
Mostly LGTM, except:

* The `WCHAR` print and comparison must make sure to work with `WCHAR` and not `wchar_t` (the latter is system dependent and might be different from `UCS2/UTF16`)
* Remove the `TestUri` parts, already done that in #8232

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@freerdp-bot test

**Review by akallabeth — APPROVED** [review]
LGTM

**Comment by akallabeth** [comment]
@KangLin I´ve fixed the remaining issues with this pr, thank you for the nice addition!

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by KangLin** [comment]
@akallabeth Why is the WCHAR definition typedef __uint16 WCHAR, not typedef wchar_t WCHAR？

**Comment by akallabeth** [comment]
@KangLin because it is the windows `WCHAR` type, which is always `16bit` whereas `wchar_t` is system dependent (most linux systems use `32bit` unicode)

**Comment by KangLin** [comment]
@akallabeth I saw the instructions in the unicode.c file.
But in general (string functions), WCHAR is defined as wchar_t. Only in FILEDESCRIPTORW need to convert to int16.
Is it too expensive to do this?
Also, if you print WCHAR?

**Comment by akallabeth** [comment]
@KangLin no, the type is used all over the place, it is core part of many protocol messages. (as `uint16`)

**Pull request merged**